clients/drcachesim/tracer/tracer.cpp

```diff
@@ -540,6 +540,13 @@ append_marker_seg_base(void *drcontext, trace_marker_type_t marker, uintptr_t va
         return; /* This thread was filtered out. */
     BUF_PTR(data->seg_base) +=
         instru->append_marker(BUF_PTR(data->seg_base), marker, value);
+    if (file_ops_func.handoff_buf == NULL) {
+        byte *buf_ptr = BUF_PTR(data->seg_base);
+        byte *redzone = data->buf_base + trace_buf_size;
+        if (buf_ptr > redzone) {
+            memtrace(drcontext, false);
```

Comment:
Are there any invariants promised by the tracer, or assumptions made by function marker readers, that no thread switch occurs in the middle of the sequence of function entry markers (id, retaddr, args)?  Are you putting a burden on readers to either split by thread up front or have to cache markers seen for resumption of the current thread?  I wonder if we should make this guarantee to simplify readers (and remove this buffer dump)?

Could you elaborate your logic behind worrying about hitting the buffer end here?  The func_pre_hook will be on function entry, which should be at the bb start.  Adding the extra marker entries there could cause the memrefs later in the bb to go further into the buffer than they would without function tracing: isn't that just as likely if not more likely than that the markers themselves, and thus this check, would cross over?  The redzone is supposed to allow a lot of slop and let us overflow and catch it afterward at the end-of-bb check.  Will adding a handful of markers really change the risk of going off the end of the redzone?

Reply:
Summary for the offline discussion:
* memtrace (or thread switch stamp) happening between consecutive of markers is bad, we should avoid it.
* risk of going off redzone still exists if markers appended before a large BB.
The approach here is to check the redzone only after appending a consecutive number of markers.

Done.